### DISCUSSION TITLE
Get class name of Authenticator that approved login in LoginSuccessEvent

### DISCUSSION THREAD

**Discussion by tip2tail** [symfony/symfony#44819] [Q&A]
Hi

Is it possible to retrieve the class of the _actual_ authenticator that was used to approve the login in `LoginSuccessEvent`?

At the moment, calling `getAuthenticator()` on the event instance simply returns a `TraceableAuthenticator`, which I can see contains the `$authenticator` property, but that is `private`.

* Is it possible to know which authenticator was used?
* If not, is it something that the Symfony maintains would support adding?  I would be happy to submit a PR for this.

Use Case: I have two different login methods, and I want to take a certain action in the LoginSuccessEvent when one of those login methods is used, but not the other.

Regards,
M

**Comment by ro0NL** [comment]
Would putting the context in eg. request attributes work? That way it becomes an implementation detail of Authenticators, rather than coupling with them.

**Comment by helmis123** [comment]
U can store the Authenticator in the passeport, and then store it in your TokenInterface class.

> public function authenticate(Request $request): Passport
>     {
>         $request->getSession()->set(Security::LAST_USERNAME, $email);
>         $rememberMeBadge =  new RememberMeBadge();
>         $rememberMeBadge->enable();
>         $passport = new Passport(
>             new UserBadge($email),
>             new PasswordCredentials($request->request->get('_password', '')),
>             [
>                 new CsrfTokenBadge('authenticate', $request->request->get('_csrf_token')),
>                 $rememberMeBadge
>             ],
>         );
>         // here
>         $passport->setAttribute('authenticator', \_\_class\_\_ );
>         return $passport;
>     }

**Comment by wouterj** [comment]
`TraceableAuthenticator` is only used when debug=true, in production this will be your real authenticator class. 

I think it's good to "unpack" the traceable authenticator in `LoginSuccessEvent::getAuthenticator()`. I'm happy to review a PR making this change on 6.1 :) For older Symfony versions, you probably need to use a workaround as proposed by @helmis123 or @ro0NL
